src/ThreeWayMerge.php

```diff
@@ -6,52 +6,316 @@
+     * @return array
+     * @throws Exception
+     */
+    protected function linesAddedOrModified(
+        array $ancestor,
+        array $local,
+        array $remote,
+        $count,
+        $count_remote,
+        $count_ancestor,
+        $count_local
+    ) {
+        $merged = [];
+        $counter = 0;
+        // For all ancestors lines, it will check local and remote
+        // and make sure only one of them has been edited.
+        // Otherwise, throw an Exception.
+        foreach ($ancestor as $key => $value) {
+            $counter = $counter + 1;
+            if ($ancestor[$key] == $local[$key]) {
+                $merged[$key] = $remote[$key];
+            } elseif ($ancestor[$key] == $remote[$key]
+                || $local[$key] == $remote[$key]) {
+                $merged[$key] = $local[$key];
+            } else {
+                throw new Exception("A conflict has occured");
+            }
+        }
+        // Once done with ancestor lines, we have hunk of
+        // lines added. We will add them as they are.
+        // If the lines are added in both remote and local,
+        // We will make sure they are same or throw exception.
+        for ($i = $count_ancestor; $i < $count; $i++) {
+            if ($count_ancestor == $count_remote) {
+                $merged[$i] == $local[$i];
+            } elseif ($count_ancestor == $count_local) {
+                $merged[$i] = $remote[$i];
+            } elseif ($count_local == $count_remote) {
+                if ($local[$i] == $remote[$i]) {
+                    $merged[$i] = $local[$i];
+                } else {
+                    throw new Exception("A conflict has occured");
+                }
+            }
+        }
+        return $merged;
+    }
+
+    /**
+     * Handles removal or modification of lines
+     *
+     * @param array $ancestor
+     * @param array $local
+     * @param array $remote
+     * @param int $count_ancestor
+     * @param int $count_local
+     * @param int $count_remote
+     *
+     * @return array
+     * @throws Exception
+     * @internal param int $count
+     *
+     */
+    protected function linesRemovedOrModified(
+        array $ancestor,
+        array $local,
+        array $remote,
+        $count_ancestor,
+        $count_local,
+        $count_remote
+    ) {
+        $merged = [];
+        $count_array = [$count_ancestor, $count_local, $count_remote];
+        sort($count_array);
+        $mincount = min($count_local, $count_ancestor, $count_remote);
+
+        // First for loop compares all 3 nodes and returns updated node.
+        for ($key = 0; $key < $mincount; $key++) {
+            if ($ancestor[$key] == $local[$key]) {
+                $merged[$key] = $remote[$key];
+            } elseif ($ancestor[$key] == $remote[$key]
+                || $local[$key] == $remote[$key]) {
+                $merged[$key] = $local[$key];
+            } else {
+                throw new Exception("A conflict has occured");
```

Comment:
Let's define and throw our own `ConflictException` class that extends the core `Exception` class.